### PR TITLE
ENH: Adds Sparse Fascicle and Gaussian Process models

### PR DESCRIPTION
**Pull request by josephmje**
Addresses #37 and #53

### CODE REVIEW

File: eddymotion/model.py [hunk 1]
```diff
@@ -53,15 +53,17 @@ def init(gtab, model="DTI", **kwargs):
...
+                    "solver": "ElasticNet",
+                    "isotropic": ExponentialIsotropicModel,
+                }
+            else:
+                from sklearn.gaussian_process import GaussianProcessRegressor
 
-            param = {
-                "isotropic": ExponentialIsotropicModel,
-            }
+                param = {"solver": GaussianProcessRegressor}
```
Comment (dPys) [inline eddymotion/model.py:66]: In the long-term, I'd bet we get more miles out of using https://docs.pymc.io/en/v3/Gaussian_Processes.html in place of sklearn's GaussianProcessRegressor but for now this seems fine!

RE: tuning this, we could start by trying to make alpha very small like what we did with the elastic-net case, but I don't know about the kernel hyperparameters. @arokem -- would the idea here be that by wrapping this in SFM, we'd avoid the need for learning the shape of a manual kernel? The default used for GP in sklearn is `ConstantKernel(1.0, constant_value_bounds="fixed" * RBF(1.0, length_scale_bounds="fixed")`.

If we're sticking with this, then I think that either way, we should remove the `fixed` option since "kernel hyperparameters are optimized during fitting unless the bounds are marked as `fixed`" and from Andersson & Sotiropoulos 2015: "The smoothness is determined from hyperparameters whose values are determined directly from the data" and "the hyperparameters will be such that the predictions are a smooth function of gradient direction and, in the case of multi-shell data, b-value." Also noteworthy was this point: "the data given by original hyperparameters is very sharp and even for a dataset with 300 points will give almost half the weight to the center-point (i.e., when predicting the signal for a diffusion direction half the information comes from the point itself)."

Comment (dPys) [reply eddymotion/model.py:66]: And for selecting the kernel, https://www.sciencedirect.com/science/article/pii/S1053811915006874?via%3Dihub#:~:text=Finding%20k(x,scaling%20or%20variance. they suggest using either exponential or spherical and then optimizing with maximum likelihood

File: eddymotion/model.py [hunk 2]
```diff
@@ -332,6 +334,59 @@ def predict(self, gradient, **kwargs):
...
+        self._solver = solver
+        if solver is None:
+            self._solver = "ElasticNet"
+
+        kwargs = {k: v for k, v in kwargs.items() if k in ("solver",)}
+        self._model = SparseFascicleModel(gtab, **kwargs)
+
+    def fit(self, data, **kwargs):
+        """Clean-up permitted args and kwargs, and call model's fit."""
+        self._model = self._model.fit(data[self._mask, ...])
```
Comment (arokem) [inline eddymotion/model.py:372]: For parallelization, we can divide the data up into chunks before calling this, and then call this separately on each chunk (on a separate thread/process)

File: docs/notebooks/Testing GP model.ipynb [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+   "execution_count": 2,
+   "id": "750f9a01-aac3-452e-8fec-0fcc4034d448",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "base_dir = Path(\"/Users/michael/projects/datasets/ds000206\")\n",
+    "bids_dir = base_dir / \"bids\"\n",
+    "derivatives_dir = base_dir / \"dmriprep\"\n",
+    "\n",
+    "dwi_file = bids_dir / \"sub-05\" / \"ses-JHU1\" / \"dwi\" / \"sub-05_ses-JHU1_acq-GD72_dwi.nii.gz\"\n",
```
Comment (oesteban) [inline docs/notebooks/Testing GP model.ipynb:44]: Does this file contain *b=0* volumes? I believe it should not (i.e., our dwi object doesn't filter them out)

File: eddymotion/model.py [hunk 4]
```diff
@@ -53,15 +53,17 @@ def init(gtab, model="DTI", **kwargs):
...
-                SFM4HMC as Model,
-                ExponentialIsotropicModel,
-            )
+        elif model.lower() in ("sfm", "gp"):
+            Model = SparseFascicleModel
+            if model.lower() == "sfm":
+                param = {
+                    "solver": "ElasticNet",
+                    "isotropic": ExponentialIsotropicModel,
+                }
```
Comment (oesteban) [inline eddymotion/model.py:62]:
```suggestion
            param = {
                "solver": "ElasticNet",
                "isotropic": ExponentialIsotropicModel,
            }
```

File: eddymotion/model.py [hunk 5]
```diff
@@ -53,15 +53,17 @@ def init(gtab, model="DTI", **kwargs):
...
+            else:
```
Comment (oesteban) [inline eddymotion/model.py:63]:
```suggestion
            if model.lower() == "gp":
```

**Review by oesteban — APPROVED** [review]

File: docs/notebooks/Testing GP model.ipynb (hunk 3 above)
Comment (josephmje) [reply docs/notebooks/Testing GP model.ipynb:44]: Yes it does. I think the load function filters out b values less than 50. 

https://github.com/nipreps/eddymotion/blob/main/eddymotion/dmri.py#L224-L226

File: eddymotion/model.py (hunk 2 above)
Comment (josephmje) [reply eddymotion/model.py:375]: Thanks @arokem, I'm going to hand it over to @dpys to work on the chunking in a separate PR

File: docs/notebooks/Testing GP model.ipynb (hunk 3 above)
Comment (josephmje) [reply docs/notebooks/Testing GP model.ipynb:44]: We just realized that the load function as it currently stands affects the AverageB0 model since no b0s will be found. Will address in a separate PR.

**Review by dPys — APPROVED** [review]
LGTM!

File: docs/notebooks/Testing GP model.ipynb (hunk 3 above)
Comment (oesteban) [reply docs/notebooks/Testing GP model.ipynb:44]: Is there an AverageB0 model? I think it's AverageDWI

Comment (oesteban) [reply docs/notebooks/Testing GP model.ipynb:44]: Meaning, it actually doesn't expect bzeros

File: eddymotion/model.py (hunk 2 above)
Comment (dPys) [reply eddymotion/model.py:373]: @arokem and @oesteban -- Did we ever decide on whether the best approach for this would be to migrate @sebastientourbier 's asyncio parallelization routines for the DTI model to a base class for parallel fit/predict that could be imported to each of the SFM and DKI API's as well? 

On a separate note-- would it be better to stick with asyncio, switch to joblib, or expose both approaches as optional?

**Comment by oesteban** [comment]
I think we should reuse Seb's solution

File: eddymotion/model.py (hunk 2 above)
Comment (arokem) [reply eddymotion/model.py:373]: I would stick with a copy of Sebastien's solution here. Let's worry about a more elegant OO design after we see that this works well here.
